packages/core-data/src/selectors.ts

```diff
@@ -265,14 +293,31 @@ export const getEntityRecord = createSelector(
 
 		const item = queriedState.items[ context ]?.[ key ];
 		if ( item && query._fields ) {
-			const filteredItem = {};
+			const filteredItem = {} as Partial< EntityRecordOf< K, N, C > >;
 			const fields = getNormalizedCommaSeparable( query._fields ) ?? [];
 			for ( let f = 0; f < fields.length; f++ ) {
 				const field = fields[ f ].split( '.' );
 				const value = get( item, field );
 				set( filteredItem, field, value );
 			}
-			return filteredItem;
+			/**
+			 * TypeScript limitation:
+			 *
+			 *    Partial< EntityRecordOf< K, N, C > >
+			 *
+			 * Is not assignable to:
+			 *
+			 *    Fields extends undefined
+			 *	    ? EntityRecordOf< K, N, C >
+			 *		  : Partial< EntityRecordOf< K, N, C > >
+			 *
+			 * At this point, even though TypeScript knows that
+			 * Fields extends undefined. Unfortunately, this forces
+			 * us to use `as any`.
+			 *
+			 * For more details, visit https://github.com/microsoft/TypeScript/issues/23132
+			 */
+			return filteredItem as any;
```

Comment:
/me looks suspiciously at this comment

going to have to think about this a little. could we get around this by supplying two function type declarations? one in which `fields` is a `string[]` and which returns `Partial<…>` and another in which `fields` is missing and which returns `EntityRecord<…>`?

Reply:
!! I like is! I just got it to work in [TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAcgrgWwEYQE4BUD29lqgXigAoBDVAcwC4A7RFVASgID4pbdUBuAKFEigDKwVAEtq5LENHiCxMlQDOwseSb5WS6eR69w0BCIAesnPSym8AMkHLxk2+SjcA9M6juAegH5u3AMaY1EpQECLAABZoAGJw1H7AIoGyAN4AvlAkClAGhjyu7p4+AUHAbHRoABKYkABmcAA29SCyoRHRsfGJ1EQAjAx5bgXe-oHBmipVtQ1NLWGRqDFxCYFEAOQkSH6r-S6DhdxAA):

```ts
type NumberToNumber = (arg:number) => number;
type StringToString = (arg:string) => string;

const eitherFunction = {} as NumberToNumber & StringToString;
const numberHopefully = eitherFunction(1);
//    ^? number
const stringHopefully = eitherFunction('abc');
//    ^? string
```

I'll try to apply this to the selector signature as well.

Reply:
@dmsnell I ended up getting nowhere with the type intersection operator. The signature overloading via the `interface` declaration as follows:

```ts
interface GetEntityRecord {
	<
		R extends EntityRecordOf< K, N >,
		C extends Context = DefaultContextOf< R >,
		K extends Kind = KindOf< R >,
		N extends Name = NameOf< R >
	>(
		state: State,
		kind: K,
		name: N,
		key: KeyOf< K, N >,
		query: EntityQuery< C >
	): Partial< EntityRecordOf< K, N, C > > | null | undefined;

	<
		R extends EntityRecordOf< K, N >,
		C extends Context = DefaultContextOf< R >,
		K extends Kind = KindOf< R >,
		N extends Name = NameOf< R >
	>(
		state: State,
		kind: K,
		name: N,
		key: KeyOf< K, N >,
		query?: Omit< EntityQuery< C >, '_fields' >
	): EntityRecordOf< K, N, C > | null | undefined;
}

const getEntityRecordImplementation: GetEntityRecord = <
	R extends EntityRecordOf< K, N >,
	C extends Context = DefaultContextOf< R >,
	K extends Kind = KindOf< R >,
	N extends Name = NameOf< R >
>(
	state: State,
	kind: K,
	name: N,
	key: KeyOf< R >,
	query?: EntityQuery< C >
) => {};
```

I don't love the proliferation of types here, but it does a good job of distinguishing between the two return types. Trade-offs, trade-offs everywhere 🤷 I've also inlined the `getEntityRecordImplementation` https://github.com/WordPress/gutenberg/pull/41235/commits/1bd4111309f7cdbb1232815ce0970b91b3c01bc8

✅ One declaration less
❌ `as GetEntityRecord` doesn't give us the same type safety as `: GetEntityRecord`